rflx/model/session.py

```diff
@@ -198,6 +198,105 @@ def _collect_message_field_assignments(
             ),
         )
 
+    def optimize(self) -> None:
+        """Execute optimization transformations on state."""
+        self._optimize_structures()
+
+    def _optimize_structures(self) -> None:
+        """
+        Replace local messages with structures if possible.
+
+        If only a limited feature set of a local message object is used it can be replaced with
+        a structure. This allows the generation of more size and runtime efficient code.
+        """
+
+        def find_identifier(name: ID, expression: expr.Expr) -> bool:
+            """Check if variable with a given identifier exists within an expression."""
+            return bool(
+                expression.findall(lambda e: isinstance(e, expr.Variable) and e.identifier == name)
+            )
+
+        def find_prefix(name: ID, expression: expr.Expr) -> bool:
+            """Check if the prefix of an expression contains a variable with a given identifier."""
+            return bool(
+                expression.findall(
+                    lambda e: isinstance(e, (expr.Attribute, expr.Opaque))
+                    and find_identifier(name, e.prefix)
+                )
+            )
+
+        def substituted(expression: expr.Expr, structure: rty.Structure) -> expr.Expr:
+            """Replace message types with structure of equal name."""
+
+            def subst(expression: expr.Expr) -> expr.Expr:
+                """
+                Substitute type of expression with structure type.
```

Comment:
How about renaming `subst` to `replace_expression_type` and removing its docstring?

Reply:
Changed.